src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperators.vb

```diff
@@ -39,6 +44,35 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return NullableValueOrDefault(operand)
         End Function
 
+        Private Shared Function AdjustIfOptimizableForConditionalBranch(operand As BoundExpression, <Out> ByRef optimizableForConditionalBranch As Boolean) As BoundExpression
+            optimizableForConditionalBranch = False
+
+            Dim current As BoundExpression = operand
+            Do
+                Select Case current.Kind
+                    Case BoundKind.BinaryOperator
+                        Dim binary = DirectCast(current, BoundBinaryOperator)
+                        If (binary.OperatorKind And BinaryOperatorKind.IsOperandOfConditionalBranch) <> 0 Then
+                            Select Case (binary.OperatorKind And BinaryOperatorKind.OpMask)
+                                Case BinaryOperatorKind.AndAlso, BinaryOperatorKind.OrElse
+                                    Debug.Assert((binary.OperatorKind And BinaryOperatorKind.Lifted) <> 0)
+                                    optimizableForConditionalBranch = True
+                                    Return binary.Update(binary.OperatorKind Or BinaryOperatorKind.OptimizableForConditionalBranch,
```

Comment:
>OptimizableForConditionalBranch [](start = 99, length = 31)

If I followed correctly, this is the only place were we set `BinaryOperatorKind.OptimizableForConditionalBranch`. It seems that all the information to decide whether to set this flag is available during binding. Why do we need to wait until lowering to set it?
If we can compute this during binding, then we would only need one flag.

Reply:
> If we can compute this during binding, then we would only need one flag.

I prefer to have stronger guarantees and apply optimization only when we are absolutely sure that we will "snap" Null to false. And that is when we actually have the ```BoundNullableIsTrueOperator``` as the ancestor and no user defined operators in between. We simply don't know that information during binding because we haven't bound binary operators up the hierarchy yet. So the optimization is triggered by the fact of snapping Null to false (getting to the VisitNullableIsTrueOperator method) and then we are making sure we don't have anything unexpected in between.

---
In reply to: [328851498](https://github.com/dotnet/roslyn/pull/38802#discussion_r328851498) [](ancestors = 328851498)

Reply:
Makes sense.
That may be a useful comment to capture next to one of the two flags.

---
In reply to: [328865595](https://github.com/dotnet/roslyn/pull/38802#discussion_r328865595) [](ancestors = 328865595,328851498)